crates/network-scanner-net/src/runtime.rs

```diff
@@ -81,130 +80,105 @@ impl Socket2Runtime {
+        self.remove_events_with_id_from_history(id);
         Ok(())
     }
 
     fn start_loop(
         &self,
         register_receiver: Receiver<RegisterEvent>,
-        event_sender: Sender<Event>,
+        event_history: Arc<DashSet<EventWrapper>>,
     ) -> anyhow::Result<()> {
-        // To prevent an Arc cycle with the Arc<Socket2Runtime>, we added additional indirection.
-        // Otherwise, the reference in the new thread would prevent the runtime from being dropped and shutdown.
-        let poller = self.poller.clone();
+        // We make is_terminated Arc<AtomicBool> and poller Arc<Poller> so that we can clone them and move them into the thread.
+        // The reason why we cannot hold a Arc<Socket2Runtime> in the thread is because it will create a cycle reference and the runtime will never be dropped.
         let is_terminated = self.is_terminated.clone();
-
+        let poller = self.poller.clone();
         std::thread::Builder::new()
             .name("[raw-socket]:io-event-loop".to_string())
             .spawn(move || {
-                let mut events = Events::with_capacity(NonZeroUsize::new(1024).unwrap());
+                let mut events = Events::with_capacity(NonZeroUsize::new(QUEUE_CAPACITY).unwrap());
                 tracing::debug!("starting io event loop");
                 // events registered but not happened yet
-                let mut events_registered = HashMap::new();
-
-                // events happened but not registered yet
-                let mut events_happened = HashMap::new();
+                let mut events_registered: HashMap<EventWrapper, Waker> = HashMap::new();
 
                 loop {
                     if is_terminated.load(Ordering::Acquire) {
                         break;
                     }
 
-                    tracing::debug!("polling events");
-                    if let Err(e) = poller.wait(&mut events, None) {
+                    // The timeout 200ms is critical, sometimes the event might be registered after the event happened
+                    // the timeout limit will allow the events to be checked periodically.
+                    if let Err(e) = poller.wait(&mut events, Some(Duration::from_millis(200))) {
                         tracing::error!(error = ?e, "failed to poll events");
                         is_terminated.store(true, Ordering::SeqCst);
                         break;
                     };
+
                     for event in events.iter() {
                         tracing::trace!(?event, "event happened");
-                        events_happened.insert(event.key, event);
+                        // This is different from just insert, as the event wrapper will have the same hash, it actually does not replace the old one.
+                        // by removing the old one first, we can make sure the new one is inserted.
+                        event_history.remove(&event.into());
```

Comment:
The underlying data structure is a hash map, and the value as set is the key to the map. Which will not be updated if the hash is the same. 
We need to replace the value in the set because we need the latest event with it's system specific event arg, in particular, for linux's OUT and HUP.
Since the underlying system specific event are not exposed, we cannot hash it, so instead, we do it this way to make sure it's the latest event that is actually in the cache,